src/base/init/init.c

```diff
@@ -422,10 +423,14 @@ void low_mem_init(void)
     ptr2 += LOWMEM_SIZE + HMASIZE;
     register_hardware_ram_virtual('U', DOSADDR_REL(ptr2), phys_rsv,
 	    LOWMEM_SIZE + HMASIZE);
-  }
-  /* create ext_mem alias for int15/dpmi */
-  register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
+    /* register ext_mem and XMS alias memory for dpmi */
+    register_hardware_ram_virtual('X', LOWMEM_SIZE + HMASIZE, phys_rsv,
 	    DOSADDR_REL(ptr2));
+  } else
+    /* without DPMI there is only registered physical memory, no alias */
+    register_hardware_ram('X', LOWMEM_SIZE + HMASIZE, phys_rsv);
```

Comment:
Wasn't that one creating the kmem
maps?

Reply:
This used to only register kmem maps, but can be used in general to define any physical region without a virtual address.

The actual kmem mapping happens in init_hardware_ram, and now checks for the type ('h' or 'v')

Reply:
OK but it doesn't call mmap_kvm().
So kmems are getting visible to kvm
when mapped to main_pool.
And here?

Reply:
here, it's just registered, but never mapped to main_pool, because it doesn't need to be mapped to main_pool any more (for $_dpmi = 0)

kmem won't map either then (because of the explicit config.dpmi check).

Reply:
I understand its not mapped to main_pool.
How it is mapped to kvm?

Reply:
it's not mapped to kvm any more either, kvm now only maps lowmem+hma and the monitor if dpmi=0. Basically the setting of config.cpu_vm_dpmi is then not relevant any more.

Reply:
But isn't that weird?
We use map_hardware_ram() for
things that _are_ mapped to kvm.
And what if vcpi maps that area?

Reply:
I mean, why not to use _virtual
as before, just pass -1 as va, and
get things mapped to kvm?

Reply:
if you pass -1 to `register_hardware_ram_virtual` it'll call
`  register_hardware_ram_virtual2(type, base, size, MEM_BASE32x(-1, KVM_BASE), -1);`
which makes no sense.

I could extend kvm_base first then, in which case something like:
`  register_hardware_ram_virtual2(type, base, size, &kvm_base[base], -1);`
could work, with appropriate adjustments in virtual2?

Reply:
Only for xms I guess?
Because for ext_mem we already have
a physical view at lowmem_base...
Yes, that sounds like a plan.

Reply:
done that as a followup patch now. If it's ok I can cleanup the history.

it's easiest to use kvm_base for everything, otherwise KVM will have multiple slots: low -> kvm_base, ext_mem -> lowmem_base, xms -> kvm_base, when alias mapping lowmem_base into kvm_base gives a single mapping for KVM.